miden-crypto/src/merkle/smt/partial/mod.rs

```diff
@@ -0,0 +1,566 @@
+        self.num_entries = self.num_entries + current_entries - prev_entries;
+
+        // Recompute the path from leaf to root
+        let new_leaf_hash = leaf.hash();
+        self.recompute_nodes_from_leaf_to_root(leaf_index, new_leaf_hash);
+
+        Ok(previous_value)
+    }
+
+    /// Adds an [`SmtProof`] to this [`PartialSmt`].
+    ///
+    /// This is a convenience method which calls [`Self::add_path`] on the proof. See its
+    /// documentation for details on errors.
+    pub fn add_proof(&mut self, proof: SmtProof) -> Result<(), MerkleError> {
+        let (path, leaf) = proof.into_parts();
+        self.add_path(leaf, path)
+    }
+
+    /// Adds a leaf and its sparse merkle path to this [`PartialSmt`].
+    ///
+    /// If this function was called, any key that is part of the `leaf` can subsequently be updated
+    /// to a new value and produce a correct new tree root.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if:
+    /// - the new root after the insertion of the leaf and the path does not match the existing
+    ///   root. If an error is returned, the tree is left in an inconsistent state.
+    pub fn add_path(&mut self, leaf: SmtLeaf, path: SparseMerklePath) -> Result<(), MerkleError> {
+        let path_root = self.add_path_unchecked(leaf, path);
+
+        // Check if the newly added merkle path is consistent with the existing tree. If not, the
+        // merkle path was invalid or computed against another tree.
+        if self.root() != path_root {
+            return Err(MerkleError::ConflictingRoots {
+                expected_root: self.root(),
+                actual_root: path_root,
+            });
+        }
+
+        Ok(())
+    }
+
+    /// Returns an iterator over the inner nodes of the [`PartialSmt`].
+    pub fn inner_nodes(&self) -> impl Iterator<Item = InnerNodeInfo> + '_ {
+        self.inner_nodes.values().map(|e| InnerNodeInfo {
+            value: e.hash(),
+            left: e.left,
+            right: e.right,
+        })
+    }
+
+    /// Returns an iterator over the [`InnerNode`] and the respective [`NodeIndex`] of the
+    /// [`PartialSmt`].
+    pub fn inner_node_indices(&self) -> impl Iterator<Item = (NodeIndex, InnerNode)> + '_ {
+        self.inner_nodes.iter().map(|(idx, inner)| (*idx, inner.clone()))
+    }
+
+    /// Returns an iterator over the explicitly stored, non-empty leaves of the [`PartialSmt`] in
+    /// arbitrary order.
+    ///
+    /// Note: This only returns leaves that were explicitly added via [`Self::add_path`] or
+    /// [`Self::add_proof`], or created through [`Self::insert`]. It does not include implicitly
+    /// trackable leaves in empty subtrees.
+    pub fn leaves(&self) -> impl Iterator<Item = (LeafIndex<SMT_DEPTH>, &SmtLeaf)> {
+        self.leaves
+            .iter()
+            .filter(|(_, leaf)| !leaf.is_empty())
+            .map(|(leaf_index, leaf)| (LeafIndex::new_max_depth(*leaf_index), leaf))
+    }
```

Comment:
Related to the previous comment: if we remove the empty leaves, we won't need to do the filtering here.